### PR TITLE
Add pass to detect combinational loops

### PR DESCRIPTION
**Pull request by albert-magyar**
Now that #479 is merged, this is more-or-less ready.

### CODE REVIEW

File: src/main/scala/firrtl/LoweringCompilers.scala [hunk 1]
```diff
@@ -97,6 +97,7 @@ class LowFirrtlOptimization extends CoreTransform {
   def inputForm = LowForm
   def outputForm = LowForm
   def passSeq = Seq(
+    passes.CheckCombLoops,
```
Comment (jackkoenig) [inline src/main/scala/firrtl/LoweringCompilers.scala]: This should not be optional

**Comment by albert-magyar** [comment]
I moved it to the end of middle-to-low compilation, as we talked about in the meeting @colinschmidt

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  * @note The pass will throw exceptions on "false paths"
+  */
+
+object CheckCombLoops extends Pass {
+  def name = "Check Loops"
+
+  class CombLoopException(info: Info, mname: String, cycle: Seq[String]) extends PassException(
+    s"$info: [module $mname] Combinational loop detected:\n" + cycle.mkString("\n"))
+
+  private case class LogicNode(name: String, inst: Option[String] = None, memport: Option[String] = None)
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: Probably worth clarifying in a comment what this is used for, and what the fields are for - I know from our previous conversation, but it isn't obvious from just looking at the class.

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      s
+    case _ =>
+      e map getExprDeps(deps)
+      e
+  }
+
+  private def getStmtDeps(
+    simplifiedModules: mutable.Map[String,DiGraph[LogicNode]],
+    deps: MutableDiGraph[LogicNode])(s: Statement): Statement = s match {
+    case c: Connect =>
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: Better to do
```
case Connect(_, loc, expr) => ...
```

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    case s: WSubField =>
+      s.exp match {
+        case modref: WRef =>
+          LogicNode(s.name,Some(modref.name))
+        case memport: WSubField =>
+          memport.exp match {
+            case memref: WRef =>
+              LogicNode(s.name,Some(memref.name),Some(memport.name))
+          }
+      }
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: add
```
case _ => throwInternalError
```

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  private def getExprDeps(deps: mutable.Set[LogicNode])(e: Expression): Expression = e match {
+    case r: WRef =>
+      deps += toLogicNode(r)
+      r
+    case s: WSubField =>
+      deps += toLogicNode(s)
+      s
+    case _ =>
+      e map getExprDeps(deps)
+      e
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: can just return `e map getExpreDeps(deps)`

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+package firrtl.passes
+
+import scala.collection.mutable
+import scala.collection.immutable.HashSet
+import scala.collection.immutable.HashMap
+import annotation.tailrec
+
+import firrtl._
+import firrtl.ir._
+import firrtl.graph._
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: expand this import

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import firrtl.Utils._
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: expand this import

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        Seq(absNodeName(prefix,a))
+      }
+    }
+    cycNodes.flatten :+ absNodeName(prefix, cycle.last)
+  }
+
+
+  def run(c: Circuit): Circuit = {
+    val errors = new Errors()
+    // TODO: deal with exmodules! No pass warnings currently exist
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: add your name (magyar) to the TODO. Doesn't mean you'll do it, just that you added this TODO

File: src/main/scala/firrtl/passes/CheckCombLoops.scala [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        recoverCycle(child,moduleGraphs,moduleDeps,newprefix,subpath)
+      } else {
+        Seq(absNodeName(prefix,a))
+      }
+    }
+    cycNodes.flatten :+ absNodeName(prefix, cycle.last)
+  }
+
+
+  def run(c: Circuit): Circuit = {
```
Comment (azidar) [inline src/main/scala/firrtl/passes/CheckCombLoops.scala]: Add an explanation of the algorithm (e.g. first topological sort, then iterate through modules, then check.

File: src/test/scala/firrtlTests/CheckCombLoopsSpec.scala [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    new IRToWorkingIR,
+    new ResolveAndCheck,
+    new HighFirrtlToMiddleFirrtl,
+    new MiddleFirrtlToLowFirrtl,
+    new TestCombLoopsCheck
+  )
+
+  "Simple combinational loop" should "throw an exception" in {
+    val input = """
+circuit hasloops :
```
Comment (azidar) [inline src/test/scala/firrtlTests/CheckCombLoopsSpec.scala]: stripMargin lets you add `|` before the text, look at other tests, e.g. UnitTests.scala. Fix this for all your tests.

**Pull request merged**